test/run_test.sh

```diff
@@ -0,0 +1,288 @@
+                TS_DISPLAY_TEST_RESULT=false
+            ;;
+            s)
+                TS_SYNC_TEST=true
+            ;;
+            t)
+                TS_TEST_TO_RUN=$OPTARG
+            ;;
+            :)
+                echo "Arg required for $OPTARG"
+                exit 1
+            ;;
+            \?)
+                echo "Invalid option: $OPTARG"
+                echo
+                help
+            ;;
+        esac
+    done
+    
+    # First args after option should be the TS_TEST_DIR
+    shift $((OPTIND-1))
+    TS_TEST_DIR=$1
+    if [[ -z $TS_TEST_DIR ]];then
+        echo "Test directory not provide"
+        help
+    fi
+    if [[ ! -d $TS_TEST_DIR ]];then
+        log_error "Test directory $TS_TEST_DIR not found"
+    fi
+    
+    # Check test files are OK
+    get_test_files > /dev/null
+}
+
+# Log
+log() {
+    echo "INFO: $@"
+}
+
+log_error() {
+    echo "ERROR: $@" >&2
+    exit 1
+}
+
+log_debug() {
+    [[ $TS_LOG_DEBUG == true ]] || return
+    echo "DEBUG: $@"
+}
+
+# This will enable log with prefix string, this method should be used in a subshell
+enable_prefix_log() {
+    local prefix=$1
+    exec > >(awk '$0="['$prefix'][INFO]\t"$0') 2> >(awk '$0="['$prefix'][ERROR]\t"$0' >&2)
+}
+# reset stdout/stderr to &1 &2
+reset_log() {
+    # restor previsously saved fd
+    exec 1>&6 2>&7
+}
+
+# Disable all output (stdout & stderr)
+disable_log() {
+    exec >/dev/null 2>/dev/null
+}
+
+# This will create all file/queue needed by this test framework
+prepare_system() {
+    TS_TMPDIR=$(mktemp -d)
+    TS_QUEUE_RESULT=$TS_TMPDIR/result
+    touch $TS_QUEUE_RESULT
+    log_debug "tmp dir: $TS_QUEUE_RESULT"
+    
+    # save filedescriptor
+    exec 6>&1 # save stdout in 6
+    exec 7>&2 # save stderr in 7
+}
+
+# Remove all file/queue created by this test framework and exit with $TS_EXIT_CODE
+shutdown_system() {
+    log_debug "shutdown framework"
+    rm -rf $TS_TMPDIR
+}
+
+# Get all test files that need to be exec
+get_test_files() {
+    if [[ ! -z $TS_TEST_TO_RUN ]]; then
+        local test_file=$TS_TEST_DIR/$TS_TEST_TO_RUN
+        [[ -e $test_file ]] || log_error "Test file not found: $test_file"
+        echo $test_file
+        return
+    fi
+    
+    ls $TS_TEST_DIR/test_* 2>&1 >/dev/null || log_error "No test file found in $TS_TEST_DIR"
+    for test_file in $TS_TEST_DIR/test_* ;do
+        echo $test_file
+    done
+}
+
+# Send test result in queue
+send_test_result() {
+    local test_name=$1
+    local test_exit_code=$2
+    echo "$2|$1" >> $TS_QUEUE_RESULT
+}
+
+# Get exist code from result string
+get_exit_code_from_result(){
+    local result=$1
+    echo $result | awk -F'|' '{ print $1}'
+}
+
+# Get test name from result string
+get_test_name_from_result(){
+    local result=$1
+    echo $result | awk -F'|' '{ print $2}'
+}
+
+# Run a single test
+run_test(){
+    local test_file=$1
+    
+    local test_name=$(basename $test_file)
+    if [[ $TS_DISPLAY_TEST_LOG == "true" ]];then
```

Comment:
Does this need to run for every test? It seems that `TS_DISPLAY_TEST_LOG` is only set once, from the command line arguments, so that redirecting output either to `/dev/null` or letting it go to stdout/stderr could be done only once, in `run_tests` or even before.

Reply:
Is not really possible because it's here I add prefix if log are not disable, and prefix is test specific ($test_name)